### DISCUSSION TITLE
SQL: How to configure with .sqlfluff

### DISCUSSION THREAD

**Discussion by embiem** [LazyVim/LazyVim#4594] [Q&A] [answered]
After installing the "SQL" LazyExtra & adding a `.sqlfluff` file at the root of my project/cwd, I receive sqlfluff errors for the following:

```sql
ALTER TABLE mytable
  ADD CONSTRAINT fk_column_id
  FOREIGN KEY (other_id)
  REFERENCES othertable (id)
  ON UPDATE RESTRICT
  ON DELETE CASCADE;
```

sqlfluff error: "Line 3, Position 3: Found unparsable section: 'FOREIGN KEY (other_id)\n  REFERENCES othe...' sqlfluff  [3, 3]"

Which I'm assuming is because of the wrong dialect. After adding this `.sqlfluff` file, I'm still receiving the errors though:
```
[sqlfluff]
dialect = postgres
```

Is there a specific way to configure sqlfluff used as linter in nvim?

I'm still pretty new to nvim, so I might have missed a configuration step.

**Comment by embiem** [comment, marked as answer]
After stumbling over this code in the nvim-lint repo: https://github.com/mfussenegger/nvim-lint/blob/962a76877a4479a535b935bd7ef35ad41ba308b2/lua/lint/linters/sqlfluff.lua#L3-L7

I have added a new file in ~/.config/nvim/lua/plugins/linter.lua with the contents:


```lua
return {
  "mfussenegger/nvim-lint",
  opts = {
    linters = {
      sqlfluff = {
        args = {
          "lint",
          "--format=json",
          -- note: users will have to replace the --dialect argument accordingly
          "--dialect=postgres",
        },
      },
    },
  },
}
```

restarted nvim and now the sqlfluff error disappeared. So it was the dialect issue.

This seems inefficient though. I'll keep searching for a way to configure sqlfluff on a project basis using the .sqlfluff file instead. For example, I wouldn't want to have to change my nvim config everytime I switch projects from a postgres dialect SQL to a sqlite dialect.

/edit: Just removing the "--dialect=postgres" argument in the above snippet fixes the issue. It is now reading the dialect correctly from my project's `.sqlfluff` file.

**Reply by wylie102** [reply]
Hi, good work on looking this up but can I check if this is still working for you? 

I have tried adding the above in a linter.lua file in plugins (with --dialect=duckdb as this is the dialect I am trying to get it to parse) but no change for me. 

I then removed that line entirely and made sure there is a .sqlfluff file in the folder with the correct information in but still no improvement. 

Was there any other step to your process? Do you have the sql lang extra enabled that bundles dadbod in? or leave that disabled? I can't seem to get it to work for me.

**Reply by talgat-ruby** [reply]
@wylie102 I am not sure about duckdb, but for postgres worked for me. Added both `linter.lua` and `.sqlfluff`. You can also check with `:ConformInfo` if sqlfluff is enabled as a formatter.

**Comment by german-enriquez** [comment]
You can also inline your linting rules inside each file following https://docs.sqlfluff.com/en/stable/configuration/setting_configuration.html#in-file-configuration-directives

**Comment by ahsanu123** [comment]
thanks @german-enriquez, work with sqlite 

```sql 
-- sqlfluff:dialect:sqlite
CREATE TABLE posts (
    id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
    title VARCHAR NOT NULL,
    body TEXT NOT NULL,
    published BOOLEAN NOT NULL DEFAULT 0
)

```
